```
Error running mint automation
+ mkdir -p 9848-c02f9ba/gopath/src/github.com/minio
+ git -C 9848-c02f9ba/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 9848-c02f9ba/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 9848-c02f9ba/gopath/src/github.com/minio/minio fetch --quiet minio pull/9848/head:pr9848
+ git -C 9848-c02f9ba/gopath/src/github.com/minio/minio checkout --quiet pr9848
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/9848-c02f9ba/gopath
+ make -C 9848-c02f9ba/gopath/src/github.com/minio/minio --quiet
go: downloading cloud.google.com/go v0.39.0
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.3.0
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/minio/minio-go/v6 v6.0.58-0.20200612001654-a57fec8037ec
go: downloading github.com/Azure/azure-pipeline-go v0.2.1
go: downloading google.golang.org/api v0.5.0
go: downloading github.com/Azure/azure-storage-blob-go v0.8.0
go: downloading github.com/montanaflynn/stats v0.5.0
go: downloading github.com/gorilla/mux v1.7.0
go: downloading github.com/gorilla/handlers v1.4.0
go: extracting github.com/gorilla/mux v1.7.0
go: extracting github.com/gorilla/handlers v1.4.0
go: extracting github.com/Azure/azure-pipeline-go v0.2.1
go: downloading github.com/coreos/etcd v3.3.12+incompatible
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: extracting github.com/montanaflynn/stats v0.5.0
go: extracting github.com/minio/cli v1.22.0
go: extracting github.com/Azure/azure-storage-blob-go v0.8.0
go: downloading github.com/prometheus/client_golang v0.9.3
go: extracting github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/klauspost/compress v1.10.3
go: extracting github.com/minio/minio-go/v6 v6.0.58-0.20200612001654-a57fec8037ec
go: downloading github.com/tinylib/msgp v1.1.1
go: extracting github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: extracting github.com/prometheus/client_golang v0.9.3
go: downloading github.com/shirou/gopsutil v2.20.3-0.20200314133625-53cec6b37e6a+incompatible
go: downloading github.com/cheggaaa/pb v1.0.28
go: extracting github.com/djherbis/atime v1.0.0
go: downloading github.com/gorilla/rpc v1.2.0
go: extracting github.com/tinylib/msgp v1.1.1
go: extracting github.com/coreos/etcd v3.3.12+incompatible
go: extracting github.com/cheggaaa/pb v1.0.28
go: downloading github.com/mattn/go-runewidth v0.0.4
go: extracting github.com/gorilla/rpc v1.2.0
go: downloading github.com/mattn/go-ieproxy v0.0.0-20190805055040-f9202b1cfdeb
go: downloading github.com/dustin/go-humanize v1.0.0
go: extracting github.com/shirou/gopsutil v2.20.3-0.20200314133625-53cec6b37e6a+incompatible
go: extracting github.com/mattn/go-runewidth v0.0.4
go: downloading github.com/philhofer/fwd v1.0.0
go: downloading github.com/colinmarc/hdfs/v2 v2.1.1
go: extracting github.com/philhofer/fwd v1.0.0
go: extracting github.com/dustin/go-humanize v1.0.0
go: downloading golang.org/x/crypto v0.0.0-20200604202706-70a84ac30bf9
go: downloading github.com/elazarl/go-bindata-assetfs v1.0.0
go: extracting cloud.google.com/go v0.39.0
go: extracting github.com/mattn/go-ieproxy v0.0.0-20190805055040-f9202b1cfdeb
go: downloading github.com/json-iterator/go v1.1.10
go: extracting github.com/colinmarc/hdfs/v2 v2.1.1
go: downloading github.com/minio/sio v0.2.0
go: downloading github.com/prometheus/common v0.4.0
go: extracting github.com/elazarl/go-bindata-assetfs v1.0.0
go: extracting github.com/json-iterator/go v1.1.10
go: extracting github.com/minio/sio v0.2.0
go: downloading github.com/coredns/coredns v1.4.0
go: extracting github.com/prometheus/common v0.4.0
go: downloading github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: extracting golang.org/x/crypto v0.0.0-20200604202706-70a84ac30bf9
go: downloading github.com/minio/sha256-simd v0.1.1
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: extracting github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: downloading gopkg.in/olivere/elastic.v5 v5.0.80
go: extracting github.com/minio/sha256-simd v0.1.1
go: downloading github.com/willf/bloom v2.0.3+incompatible
go: extracting github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: extracting github.com/willf/bloom v2.0.3+incompatible
go: downloading google.golang.org/grpc v1.22.0
go: extracting gopkg.in/olivere/elastic.v5 v5.0.80
go: extracting github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: extracting github.com/coredns/coredns v1.4.0
go: extracting github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: downloading github.com/miekg/dns v1.1.8
go: downloading github.com/willf/bitset v1.1.10
go: extracting github.com/miekg/dns v1.1.8
go: downloading github.com/modern-go/reflect2 v1.0.1
go: extracting github.com/willf/bitset v1.1.10
go: downloading github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: extracting github.com/modern-go/reflect2 v1.0.1
go: downloading golang.org/x/sys v0.0.0-20200610111108-226ff32320da
go: extracting github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: extracting github.com/matttproud/golang_protobuf_extensions v1.0.1
go: extracting google.golang.org/grpc v1.22.0
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/Shopify/sarama v1.24.1
go: extracting github.com/klauspost/compress v1.10.3
go: downloading github.com/nats-io/stan.go v0.4.5
go: downloading github.com/googleapis/gax-go/v2 v2.0.4
go: extracting github.com/ncw/directio v1.0.5
go: extracting github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading google.golang.org/genproto v0.0.0-20190513181449-d00d292a067c
go: downloading github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: extracting github.com/nats-io/stan.go v0.4.5
go: downloading github.com/golang/protobuf v1.3.2
go: extracting golang.org/x/sys v0.0.0-20200610111108-226ff32320da
go: extracting github.com/googleapis/gax-go/v2 v2.0.4
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: extracting github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: downloading github.com/nats-io/nuid v1.0.1
go: extracting github.com/Shopify/sarama v1.24.1
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/klauspost/reedsolomon v1.9.9
go: extracting github.com/nats-io/nuid v1.0.1
go: extracting github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/rs/cors v1.7.0
go: extracting github.com/fatih/structs v1.1.0
go: downloading github.com/minio/simdjson-go v0.1.5-0.20200303142138-b17fe061ea37
go: extracting github.com/golang/protobuf v1.3.2
go: downloading github.com/alecthomas/participle v0.2.1
go: extracting github.com/rs/cors v1.7.0
go: downloading github.com/pkg/errors v0.8.1
go: extracting github.com/alecthomas/participle v0.2.1
go: downloading github.com/klauspost/pgzip v1.2.1
go: extracting github.com/minio/simdjson-go v0.1.5-0.20200303142138-b17fe061ea37
go: extracting github.com/klauspost/reedsolomon v1.9.9
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: extracting github.com/pkg/errors v0.8.1
go: extracting github.com/klauspost/pgzip v1.2.1
go: downloading github.com/klauspost/readahead v1.3.1
go: extracting github.com/klauspost/readahead v1.3.1
go: downloading github.com/minio/highwayhash v1.0.0
go: extracting github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/gogo/protobuf v1.2.1
go: extracting github.com/minio/highwayhash v1.0.0
go: downloading golang.org/x/net v0.0.0-20200602114024-627f9648deb9
go: extracting google.golang.org/genproto v0.0.0-20190513181449-d00d292a067c
go: downloading github.com/spaolacci/murmur3 v1.1.0
go: downloading github.com/google/uuid v1.1.1
go: extracting golang.org/x/net v0.0.0-20200602114024-627f9648deb9
go: downloading gopkg.in/ini.v1 v1.57.0
go: extracting github.com/spaolacci/murmur3 v1.1.0
go: extracting github.com/google/uuid v1.1.1
go: downloading github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/mattn/go-isatty v0.0.8
go: extracting gopkg.in/ini.v1 v1.57.0
go: downloading github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: extracting github.com/nsqio/go-nsq v1.0.7
go: extracting github.com/mattn/go-isatty v0.0.8
go: downloading github.com/fatih/color v1.7.0
go: downloading github.com/golang/snappy v0.0.1
go: extracting github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: downloading github.com/mattn/go-colorable v0.1.4
go: extracting github.com/fatih/color v1.7.0
go: downloading github.com/lib/pq v1.1.1
go: extracting github.com/mattn/go-colorable v0.1.4
go: extracting github.com/golang/snappy v0.0.1
go: downloading github.com/beevik/ntp v0.2.0
go: extracting github.com/gogo/protobuf v1.2.1
go: extracting github.com/lib/pq v1.1.1
go: extracting github.com/beevik/ntp v0.2.0
go: extracting gopkg.in/jcmturner/gokrb5.v7 v7.3.0
go: downloading github.com/minio/parquet-go v0.0.0-20200414234858-838cfa8aae61
go: downloading github.com/hashicorp/vault/api v1.0.4
go: downloading github.com/hashicorp/go-uuid v1.0.1
go: downloading go.opencensus.io v0.21.0
go: downloading github.com/klauspost/cpuid v1.3.0
go: downloading github.com/minio/md5-simd v1.1.0
go: downloading github.com/nats-io/nats.go v1.9.1
go: extracting github.com/minio/parquet-go v0.0.0-20200414234858-838cfa8aae61
go: extracting github.com/hashicorp/go-uuid v1.0.1
go: extracting google.golang.org/api v0.5.0
go: downloading github.com/beorn7/perks v1.0.0
go: extracting github.com/hashicorp/vault/api v1.0.4
go: downloading github.com/hashicorp/vault/sdk v0.1.13
go: extracting github.com/nats-io/nats.go v1.9.1
go: extracting github.com/minio/md5-simd v1.1.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: downloading github.com/hashicorp/errwrap v1.0.0
go: extracting github.com/beorn7/perks v1.0.0
go: downloading github.com/cespare/xxhash v1.1.0
go: extracting github.com/klauspost/cpuid v1.3.0
go: extracting github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: downloading github.com/eapache/go-resiliency v1.2.0
go: extracting github.com/hashicorp/vault/sdk v0.1.13
go: extracting github.com/hashicorp/errwrap v1.0.0
go: extracting github.com/cespare/xxhash v1.1.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.2.0
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: extracting go.opencensus.io v0.21.0
go: extracting github.com/eapache/go-resiliency v1.2.0
go: downloading gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: downloading github.com/mitchellh/mapstructure v1.1.2
go: extracting github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/hashicorp/go-cleanhttp v0.5.1
go: extracting github.com/eclipse/paho.mqtt.golang v1.2.0
go: extracting github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/pierrec/lz4 v2.4.0+incompatible
go: extracting gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: downloading github.com/prometheus/procfs v0.0.0-20190507164030-5867b95ac084
go: extracting github.com/mitchellh/mapstructure v1.1.2
go: downloading github.com/hashicorp/hcl v1.0.0
go: extracting github.com/hashicorp/go-cleanhttp v0.5.1
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: extracting github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading golang.org/x/text v0.3.2
go: extracting github.com/prometheus/procfs v0.0.0-20190507164030-5867b95ac084
go: extracting github.com/hashicorp/hcl v1.0.0
go: downloading github.com/hashicorp/go-retryablehttp v0.5.4
go: downloading github.com/hashicorp/golang-lru v0.5.1
go: downloading github.com/nats-io/jwt v0.3.2
go: downloading github.com/nats-io/nkeys v0.1.3
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: extracting github.com/hashicorp/golang-lru v0.5.1
go: extracting github.com/hashicorp/go-retryablehttp v0.5.4
go: downloading golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
go: downloading github.com/hashicorp/go-rootcerts v1.0.1
go: extracting gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading github.com/xdg/stringprep v1.0.0
go: extracting github.com/hashicorp/go-rootcerts v1.0.1
go: extracting golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
go: downloading gopkg.in/ldap.v3 v3.0.3
go: extracting github.com/xdg/stringprep v1.0.0
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: extracting gopkg.in/ldap.v3 v3.0.3
go: downloading gopkg.in/yaml.v2 v2.2.8
go: extracting github.com/nats-io/nkeys v0.1.3
go: extracting github.com/nats-io/jwt v0.3.2
go: extracting github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: extracting gopkg.in/yaml.v2 v2.2.8
go: downloading gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: extracting github.com/eapache/queue v1.1.0
go: extracting github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading github.com/tidwall/gjson v1.3.5
go: extracting github.com/rjeczalik/notify v0.9.2
go: extracting gopkg.in/square/go-jose.v2 v2.3.1
go: extracting gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading github.com/tidwall/sjson v1.0.4
go: extracting gopkg.in/jcmturner/aescts.v1 v1.0.1
go: extracting github.com/hashicorp/go-sockaddr v1.0.2
go: extracting github.com/tidwall/gjson v1.3.5
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading github.com/secure-io/sio-go v0.3.0
go: extracting github.com/ryanuber/go-glob v1.0.0
go: downloading golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: downloading github.com/tidwall/pretty v1.0.0
go: extracting github.com/tidwall/sjson v1.0.4
go: downloading github.com/tidwall/match v1.0.1
go: extracting gopkg.in/jcmturner/rpc.v1 v1.1.0
go: extracting git.apache.org/thrift.git v0.13.0
go: extracting golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: extracting github.com/secure-io/sio-go v0.3.0
go: extracting github.com/tidwall/match v1.0.1
go: extracting github.com/tidwall/pretty v1.0.0
go: extracting golang.org/x/text v0.3.2
go: extracting github.com/pierrec/lz4 v2.4.0+incompatible
go: finding github.com/colinmarc/hdfs/v2 v2.1.1
go: finding gopkg.in/jcmturner/gokrb5.v7 v7.3.0
go: finding github.com/minio/cli v1.22.0
go: finding github.com/minio/minio-go/v6 v6.0.58-0.20200612001654-a57fec8037ec
go: finding github.com/Azure/azure-pipeline-go v0.2.1
go: finding github.com/minio/sha256-simd v0.1.1
go: finding google.golang.org/api v0.5.0
go: finding github.com/dustin/go-humanize v1.0.0
go: finding cloud.google.com/go v0.39.0
go: finding github.com/json-iterator/go v1.1.10
go: finding github.com/minio/highwayhash v1.0.0
go: finding github.com/fatih/color v1.7.0
go: finding github.com/Azure/azure-storage-blob-go v0.8.0
go: finding github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: finding github.com/mattn/go-isatty v0.0.8
go: finding github.com/mattn/go-colorable v0.1.4
go: finding github.com/gorilla/mux v1.7.0
go: finding github.com/montanaflynn/stats v0.5.0
go: finding github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: finding github.com/rjeczalik/notify v0.9.2
go: finding github.com/secure-io/sio-go v0.3.0
go: finding github.com/shirou/gopsutil v2.20.3-0.20200314133625-53cec6b37e6a+incompatible
go: finding golang.org/x/crypto v0.0.0-20200604202706-70a84ac30bf9
go: finding golang.org/x/net v0.0.0-20200602114024-627f9648deb9
go: finding github.com/ncw/directio v1.0.5
go: finding github.com/golang/protobuf v1.3.2
go: finding github.com/coreos/etcd v3.3.12+incompatible
go: finding github.com/cespare/xxhash/v2 v2.1.1
go: finding github.com/elazarl/go-bindata-assetfs v1.0.0
go: finding golang.org/x/sys v0.0.0-20200610111108-226ff32320da
go: finding github.com/djherbis/atime v1.0.0
go: finding github.com/cheggaaa/pb v1.0.28
go: finding github.com/google/uuid v1.1.1
go: finding github.com/gorilla/handlers v1.4.0
go: finding github.com/gorilla/rpc v1.2.0
go: finding github.com/hashicorp/go-uuid v1.0.1
go: finding github.com/klauspost/readahead v1.3.1
go: finding github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: finding github.com/klauspost/compress v1.10.3
go: finding github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: finding gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: finding gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: finding github.com/klauspost/reedsolomon v1.9.9
go: finding github.com/mitchellh/go-homedir v1.1.0
go: finding gopkg.in/ini.v1 v1.57.0
go: finding github.com/minio/md5-simd v1.1.0
go: finding gopkg.in/jcmturner/aescts.v1 v1.0.1
go: finding github.com/mattn/go-ieproxy v0.0.0-20190805055040-f9202b1cfdeb
go: finding github.com/coredns/coredns v1.4.0
go: finding golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
go: finding github.com/miekg/dns v1.1.8
go: finding gopkg.in/ldap.v3 v3.0.3
go: finding google.golang.org/grpc v1.22.0
go: finding google.golang.org/genproto v0.0.0-20190513181449-d00d292a067c
go: finding github.com/googleapis/gax-go/v2 v2.0.4
go: finding go.opencensus.io v0.21.0
go: finding github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: finding github.com/modern-go/reflect2 v1.0.1
go: finding github.com/hashicorp/vault/api v1.0.4
go: finding github.com/beevik/ntp v0.2.0
go: finding github.com/Shopify/sarama v1.24.1
go: finding github.com/minio/sio v0.2.0
go: finding github.com/go-sql-driver/mysql v1.5.0
go: finding gopkg.in/jcmturner/rpc.v1 v1.1.0
go: finding github.com/eclipse/paho.mqtt.golang v1.2.0
go: finding github.com/gomodule/redigo v2.0.0+incompatible
go: finding github.com/lib/pq v1.1.1
go: finding golang.org/x/text v0.3.2
go: finding github.com/nats-io/nats.go v1.9.1
go: finding github.com/gogo/protobuf v1.2.1
go: finding github.com/nats-io/stan.go v0.4.5
go: finding github.com/pkg/errors v0.8.1
go: finding github.com/nsqio/go-nsq v1.0.7
go: finding github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: finding github.com/fatih/structs v1.1.0
go: finding github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: finding github.com/mattn/go-runewidth v0.0.4
go: finding gopkg.in/yaml.v2 v2.2.8
go: finding github.com/klauspost/pgzip v1.2.1
go: finding github.com/prometheus/client_golang v0.9.3
go: finding github.com/rs/cors v1.7.0
go: finding github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: finding github.com/willf/bloom v2.0.3+incompatible
go: finding github.com/tinylib/msgp v1.1.1
go: finding github.com/minio/parquet-go v0.0.0-20200414234858-838cfa8aae61
go: finding github.com/minio/simdjson-go v0.1.5-0.20200303142138-b17fe061ea37
go: finding github.com/alecthomas/participle v0.2.1
go: finding github.com/klauspost/cpuid v1.3.0
go: finding gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: finding gopkg.in/olivere/elastic.v5 v5.0.80
go: finding github.com/hashicorp/golang-lru v0.5.1
go: finding github.com/hashicorp/errwrap v1.0.0
go: finding github.com/hashicorp/go-retryablehttp v0.5.4
go: finding github.com/hashicorp/go-cleanhttp v0.5.1
go: finding github.com/hashicorp/vault/sdk v0.1.13
go: finding github.com/hashicorp/hcl v1.0.0
go: finding golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: finding github.com/hashicorp/go-multierror v1.0.0
go: finding github.com/hashicorp/go-rootcerts v1.0.1
go: finding github.com/mitchellh/mapstructure v1.1.2
go: finding gopkg.in/square/go-jose.v2 v2.3.1
go: finding github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: finding github.com/eapache/queue v1.1.0
go: finding github.com/eapache/go-resiliency v1.2.0
go: finding github.com/pierrec/lz4 v2.4.0+incompatible
go: finding github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: finding github.com/davecgh/go-spew v1.1.1
go: finding github.com/nats-io/nuid v1.0.1
go: finding github.com/nats-io/nkeys v0.1.3
go: finding github.com/nats-io/jwt v0.3.2
go: finding github.com/golang/snappy v0.0.1
go: finding github.com/xdg/stringprep v1.0.0
go: finding github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: finding github.com/beorn7/perks v1.0.0
go: finding github.com/prometheus/common v0.4.0
go: finding github.com/prometheus/procfs v0.0.0-20190507164030-5867b95ac084
go: finding github.com/spaolacci/murmur3 v1.1.0
go: finding github.com/willf/bitset v1.1.10
go: finding github.com/philhofer/fwd v1.0.0
go: finding github.com/tidwall/sjson v1.0.4
go: finding github.com/tidwall/gjson v1.3.5
go: finding git.apache.org/thrift.git v0.13.0
go: finding github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: finding github.com/ryanuber/go-glob v1.0.0
go: finding github.com/hashicorp/go-sockaddr v1.0.2
go: finding github.com/matttproud/golang_protobuf_extensions v1.0.1
go: finding github.com/tidwall/match v1.0.1
go: finding github.com/tidwall/pretty v1.0.0
# github.com/minio/minio/cmd
cmd/bucket-handlers.go:993:36: undefined: getBucketVersioningResponse
make: *** [build] Error 2

```

--- a/cmd/merge-walk-pool.go
+++ b/cmd/merge-walk-pool.go
@@ -35,6 +35,7 @@ type mergeWalkVersions struct {
 
 // mergeWalk - represents the go routine that does the merge walk.
 type mergeWalk struct {
+	added      time.Time
 	entryChs   []FileInfoCh
 	endWalkCh  chan struct{}   // To signal when mergeWalk go-routine should end.
 	endTimerCh chan<- struct{} // To signal when timer go-routine should end.
@@ -160,7 +161,7 @@ func NewMergeWalkPool(timeout time.Duration) *MergeWalkPool {
 // Release - selects a mergeWalk from the pool based on the input
 // listParams, removes it from the pool, and returns the MergeWalkResult
 // channel.
-// Returns nil if listParams does not have an asccociated mergeWalk.
+// Returns nil if listParams does not have an associated mergeWalk.
 func (t *MergeWalkPool) Release(params listParams) ([]FileInfoCh, chan struct{}) {
 	t.Lock()
 	defer t.Unlock()
@@ -169,6 +170,7 @@ func (t *MergeWalkPool) Release(params listParams) ([]FileInfoCh, chan struct{})
 		if len(walks) > 0 {
 			// Pop out the first valid walk entry.
 			walk := walks[0]
+			walks[0] = mergeWalk{} // clear references.
 			walks = walks[1:]
 			if len(walks) > 0 {
 				t.pool[params] = walks
@@ -195,17 +197,53 @@ func (t *MergeWalkPool) Set(params listParams, resultChs []FileInfoCh, endWalkCh
 	t.Lock()
 	defer t.Unlock()
 
+	// If we are above the limit delete at least one entry from the pool.
+	if len(t.pool) > treeWalkEntryLimit {
+		age := time.Now()
+		var oldest listParams
+		for k, v := range t.pool {
+			if len(v) == 0 {
+				delete(t.pool, k)
+				continue
+			}
+			if v[0].added.Before(age) {
+				oldest = k
+			}
+		}
+		// Invalidate and delete oldest.
+		if walks, ok := t.pool[oldest]; ok {
+			walk := walks[0]
+			walks[0] = mergeWalk{} // clear references.
+			walks = walks[1:]
+			if len(walks) > 0 {
+				t.pool[params] = walks
+			} else {
+				delete(t.pool, params)
+			}
+			walk.endTimerCh <- struct{}{}
+		}
+	}
+
 	// Should be a buffered channel so that Release() never blocks.
 	endTimerCh := make(chan struct{}, 1)
 
 	walkInfo := mergeWalk{
+		added:      UTCNow(),
 		entryChs:   resultChs,
 		endWalkCh:  endWalkCh,
 		endTimerCh: endTimerCh,
 	}
 
 	// Append new walk info.
-	t.pool[params] = append(t.pool[params], walkInfo)
+	walks := t.pool[params]
+	if len(walks) < treeWalkSameEntryLimit {
+		t.pool[params] = append(walks, walkInfo)
+	} else {
+		// We are at limit, invalidate oldest, move list down and add new as last.
+		walks[0].endTimerCh <- struct{}{}
+		copy(walks, walks[1:])
+		walks[len(walks)-1] = walkInfo
+	}
 
 	// Timer go-routine which times out after t.timeOut seconds.
 	go func(endTimerCh <-chan struct{}, walkInfo mergeWalk) {

--- a/cmd/merge-walk-pool.go
+++ b/cmd/merge-walk-pool.go
@@ -206,6 +206,7 @@ func (t *MergeWalkPool) Set(params listParams, resultChs []FileInfoCh, endWalkCh
 				delete(t.pool, k)
 				continue
 			}
+			// The first element is the oldest, so we only check that.
 			if v[0].added.Before(age) {
 				oldest = k
 			}

--- a/cmd/merge-walk-pool_test.go
+++ b/cmd/merge-walk-pool_test.go
@@ -74,7 +74,7 @@ func TestManyMergeWalksSameParam(t *testing.T) {
 		break
 	default:
 		// Create many treeWalk go-routines for the same params.
-		for i := 0; i < 10; i++ {
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
 			endWalkCh := make(chan struct{})
 			walkChs := make([]FileInfoCh, 0)
 			tw.Set(params, walkChs, endWalkCh)
@@ -82,16 +82,62 @@ func TestManyMergeWalksSameParam(t *testing.T) {
 
 		tw.Lock()
 		if walks, ok := tw.pool[params]; ok {
-			if len(walks) != 10 {
+			if len(walks) != treeWalkSameEntryLimit {
 				t.Error("There aren't as many walks as were Set")
 			}
 		}
 		tw.Unlock()
-		for i := 0; i < 10; i++ {
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
 			tw.Lock()
 			if walks, ok := tw.pool[params]; ok {
-				// Before ith Release we should have 10-i treeWalk go-routines.
-				if 10-i != len(walks) {
+				// Before ith Release we should have n-i treeWalk go-routines.
+				if treeWalkSameEntryLimit-i != len(walks) {
+					t.Error("There aren't as many walks as were Set")
+				}
+			}
+			tw.Unlock()
+			tw.Release(params)
+		}
+	}
+
+}
+
+// Test if multiple merge walkers for the same listParams are managed as expected by the pool
+// but that treeWalkSameEntryLimit is respected.
+func TestManyMergeWalksSameParamPrune(t *testing.T) {
+	// Create a treeWalkPool.
+	tw := NewMergeWalkPool(5 * time.Second)
+
+	// Create sample params.
+	params := listParams{
+		bucket: "test-bucket",
+	}
+
+	select {
+	// This timeout is an upper-bound. This is started
+	// before the first treeWalk go-routine's timeout period starts.
+	case <-time.After(5 * time.Second):
+		break
+	default:
+		// Create many treeWalk go-routines for the same params.
+		for i := 0; i < treeWalkSameEntryLimit*4; i++ {
+			endWalkCh := make(chan struct{})
+			walkChs := make([]FileInfoCh, 0)
+			tw.Set(params, walkChs, endWalkCh)
+		}
+
+		tw.Lock()
+		if walks, ok := tw.pool[params]; ok {
+			if len(walks) > treeWalkSameEntryLimit {
+				t.Error("There aren't as many walks as were Set")
+			}
+		}
+		tw.Unlock()
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
+			tw.Lock()
+			if walks, ok := tw.pool[params]; ok {
+				// Before ith Release we should have n-i treeWalk go-routines.
+				if treeWalkSameEntryLimit-i != len(walks) {
 					t.Error("There aren't as many walks as were Set")
 				}
 			}

--- a/cmd/tree-walk-pool.go
+++ b/cmd/tree-walk-pool.go
@@ -120,6 +120,7 @@ func (t *TreeWalkPool) Set(params listParams, resultCh chan TreeWalkResult, endW
 				delete(t.pool, k)
 				continue
 			}
+			// The first element is the oldest, so we only check that.
 			if v[0].added.Before(age) {
 				oldest = k
 			}

--- a/cmd/tree-walk-pool.go
+++ b/cmd/tree-walk-pool.go
@@ -25,7 +25,9 @@ import (
 
 // Global lookup timeout.
 const (
-	globalLookupTimeout = time.Minute * 30 // 30minutes.
+	globalLookupTimeout    = time.Minute * 30 // 30minutes.
+	treeWalkEntryLimit     = 50
+	treeWalkSameEntryLimit = 4
 )
 
 // listParams - list object params used for list object map
@@ -44,6 +46,7 @@ var errWalkAbort = errors.New("treeWalk abort")
 
 // treeWalk - represents the go routine that does the file tree walk.
 type treeWalk struct {
+	added      time.Time
 	resultCh   chan TreeWalkResult
 	endWalkCh  chan struct{}   // To signal when treeWalk go-routine should end.
 	endTimerCh chan<- struct{} // To signal when timer go-routine should end.
@@ -72,7 +75,7 @@ func NewTreeWalkPool(timeout time.Duration) *TreeWalkPool {
 // Release - selects a treeWalk from the pool based on the input
 // listParams, removes it from the pool, and returns the TreeWalkResult
 // channel.
-// Returns nil if listParams does not have an asccociated treeWalk.
+// Returns nil if listParams does not have an associated treeWalk.
 func (t *TreeWalkPool) Release(params listParams) (resultCh chan TreeWalkResult, endWalkCh chan struct{}) {
 	t.Lock()
 	defer t.Unlock()
@@ -81,6 +84,7 @@ func (t *TreeWalkPool) Release(params listParams) (resultCh chan TreeWalkResult,
 		if len(walks) > 0 {
 			// Pop out the first valid walk entry.
 			walk := walks[0]
+			walks[0] = treeWalk{} // clear references.
 			walks = walks[1:]
 			if len(walks) > 0 {
 				t.pool[params] = walks
@@ -100,22 +104,58 @@ func (t *TreeWalkPool) Release(params listParams) (resultCh chan TreeWalkResult,
 // 1) time.After() expires after t.timeOut seconds.
 //    The expiration is needed so that the treeWalk go-routine resources are freed after a timeout
 //    if the S3 client does only partial listing of objects.
-// 2) Relase() signals the timer go-routine to end on endTimerCh.
+// 2) Release() signals the timer go-routine to end on endTimerCh.
 //    During listing the timer should not timeout and end the treeWalk go-routine, hence the
 //    timer go-routine should be ended.
 func (t *TreeWalkPool) Set(params listParams, resultCh chan TreeWalkResult, endWalkCh chan struct{}) {
 	t.Lock()
 	defer t.Unlock()
 
+	// If we are above the limit delete at least one entry from the pool.
+	if len(t.pool) > treeWalkEntryLimit {
+		age := time.Now()
+		var oldest listParams
+		for k, v := range t.pool {
+			if len(v) == 0 {
+				delete(t.pool, k)
+				continue
+			}
+			if v[0].added.Before(age) {
+				oldest = k
+			}
+		}
+		// Invalidate and delete oldest.
+		if walks, ok := t.pool[oldest]; ok {
+			walk := walks[0]
+			walks[0] = treeWalk{} // clear references.
+			walks = walks[1:]
+			if len(walks) > 0 {
+				t.pool[params] = walks
+			} else {
+				delete(t.pool, params)
+			}
+			walk.endTimerCh <- struct{}{}
+		}
+	}
+
 	// Should be a buffered channel so that Release() never blocks.
 	endTimerCh := make(chan struct{}, 1)
 	walkInfo := treeWalk{
+		added:      UTCNow(),
 		resultCh:   resultCh,
 		endWalkCh:  endWalkCh,
 		endTimerCh: endTimerCh,
 	}
 	// Append new walk info.
-	t.pool[params] = append(t.pool[params], walkInfo)
+	walks := t.pool[params]
+	if len(walks) < treeWalkSameEntryLimit {
+		t.pool[params] = append(walks, walkInfo)
+	} else {
+		// We are at limit, invalidate oldest, move list down and add new as last.
+		walks[0].endTimerCh <- struct{}{}
+		copy(walks, walks[1:])
+		walks[len(walks)-1] = walkInfo
+	}
 
 	// Timer go-routine which times out after t.timeOut seconds.
 	go func(endTimerCh <-chan struct{}) {

--- a/cmd/tree-walk-pool_test.go
+++ b/cmd/tree-walk-pool_test.go
@@ -74,7 +74,7 @@ func TestManyWalksSameParam(t *testing.T) {
 		break
 	default:
 		// Create many treeWalk go-routines for the same params.
-		for i := 0; i < 10; i++ {
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
 			resultCh := make(chan TreeWalkResult)
 			endWalkCh := make(chan struct{})
 			tw.Set(params, resultCh, endWalkCh)
@@ -82,16 +82,16 @@ func TestManyWalksSameParam(t *testing.T) {
 
 		tw.Lock()
 		if walks, ok := tw.pool[params]; ok {
-			if len(walks) != 10 {
+			if len(walks) != treeWalkSameEntryLimit {
 				t.Error("There aren't as many walks as were Set")
 			}
 		}
 		tw.Unlock()
-		for i := 0; i < 10; i++ {
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
 			tw.Lock()
 			if walks, ok := tw.pool[params]; ok {
-				// Before ith Release we should have 10-i treeWalk go-routines.
-				if 10-i != len(walks) {
+				// Before ith Release we should have n-i treeWalk go-routines.
+				if treeWalkSameEntryLimit-i != len(walks) {
 					t.Error("There aren't as many walks as were Set")
 				}
 			}
@@ -99,5 +99,49 @@ func TestManyWalksSameParam(t *testing.T) {
 			tw.Release(params)
 		}
 	}
+}
 
+// Test if multiple tree walkers for the same listParams are managed as expected by the pool
+// but that treeWalkSameEntryLimit is respected.
+func TestManyWalksSameParamPrune(t *testing.T) {
+	// Create a treeWalkPool.
+	tw := NewTreeWalkPool(5 * time.Second)
+
+	// Create sample params.
+	params := listParams{
+		bucket: "test-bucket",
+	}
+
+	select {
+	// This timeout is an upper-bound. This is started
+	// before the first treeWalk go-routine's timeout period starts.
+	case <-time.After(5 * time.Second):
+		break
+	default:
+		// Create many treeWalk go-routines for the same params.
+		for i := 0; i < treeWalkSameEntryLimit*4; i++ {
+			resultCh := make(chan TreeWalkResult)
+			endWalkCh := make(chan struct{})
+			tw.Set(params, resultCh, endWalkCh)
+		}
+
+		tw.Lock()
+		if walks, ok := tw.pool[params]; ok {
+			if len(walks) != treeWalkSameEntryLimit {
+				t.Error("There aren't as many walks as were Set")
+			}
+		}
+		tw.Unlock()
+		for i := 0; i < treeWalkSameEntryLimit; i++ {
+			tw.Lock()
+			if walks, ok := tw.pool[params]; ok {
+				// Before ith Release we should have n-i treeWalk go-routines.
+				if treeWalkSameEntryLimit-i != len(walks) {
+					t.Error("There aren't as many walks as were Set")
+				}
+			}
+			tw.Unlock()
+			tw.Release(params)
+		}
+	}
 }

